File: 프로그래머스/2/87946. 피로도/피로도.py
answer = 0
visited = []
n = 0

def dfs(k, cnt, dungeons):
    global answer, visited,n
    if cnt > answer:
        answer = cnt
    for i in range(n):
        if k >= dungeons[i][0] and not visited[i]:
            visited[i] = 1
            dfs(k - dungeons[i][1], cnt + 1, dungeons)
            visited[i] = 0
        

def solution(k, dungeons):
    global visited,n
    n = len(dungeons)
    visited = [0] * n
    dfs(k, 0, dungeons)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <bits/stdc++.h>

using namespace std;

int ret, visited[10];

void backtracking(int k, vector<vector<int>> dungeons,  int cnt) {
    ret = max(ret, cnt);
    
    for (int i=0;i<dungeons.size();i++) {
        if (!visited[i] && k >= dungeons[i][0]) {
            visited[i] = 1;
            backtracking(k - dungeons[i][1], dungeons, cnt + 1);
            visited[i] = 0;
        }
    }
    
}


int solution(int k, vector<vector<int>> dungeons) {
    
    backtracking(k, dungeons, 0);
    
    return ret;
}